### ISSUE TITLE
feat: implement early stopping for training

### ISSUE THREAD

**Issue by aymuos15** [aymuos15/nnBenchmark#17]
## Feature Request

Implement early stopping to automatically halt training when validation performance stops improving, preventing overfitting and reducing unnecessary computation.

## Current Behavior

Training runs for a fixed number of epochs specified in the config (e.g., `epochs: 1000`), regardless of whether the model is still improving. This can lead to:
- **Overfitting**: Model continues training after optimal validation performance
- **Wasted resources**: Training continues unnecessarily after convergence
- **Manual monitoring**: Users must manually monitor training and stop it

## Proposed Behavior

Add configurable early stopping that monitors validation metrics and stops training when performance plateaus.

### CLI Usage
```bash
# Early stopping enabled via config
nnBench.train --config fold_0.yaml --dataset Dataset001_Cellpose
```

### Configuration
```yaml
training:
  epochs: 1000
  early_stopping:
    enabled: true
    monitor: DiceMetric           # Metric to monitor
    patience: 50                  # Epochs to wait without improvement
    min_delta: 0.0001            # Minimum change to qualify as improvement
    mode: max                     # max (for metrics) or min (for losses)
    restore_best_weights: true    # Load best checkpoint after stopping
```

## Use Cases

1. **Prevent overfitting**: Stop when validation performance degrades
2. **Save computation**: Halt training early when model has converged
3. **Automatic tuning**: Find optimal training duration without manual intervention
4. **Resource efficiency**: Free up GPUs faster for other experiments
5. **Reproducibility**: Consistent stopping criteria across runs

### Core Logic
```python
class EarlyStopping:
    def __init__(self, patience=50, min_delta=0.0001, mode='max'):
        self.patience = patience
        self.min_delta = min_delta
        self.mode = mode
        self.counter = 0
        self.best_score = None
        self.should_stop = False
        
    def __call__(self, current_score):
        if self.best_score is None:
            self.best_score = current_score
        elif self._is_improvement(current_score):
            self.best_score = current_score
            self.counter = 0
        else:
            self.counter += 1
            if self.counter >= self.patience:
                self.should_stop = True
```

### Integration Points

1. **Handler**: Create `EarlyStoppingHandler` in Ignite framework
2. **Config**: Add early_stopping section to training config
3. **Logging**: Record when early stopping is triggered
4. **Checkpointing**: Optionally restore best checkpoint on stop
5. **History**: Save early stopping info in `training_history.json`

### Output Example
```
Epoch 450/1000 - DiceMetric: 0.8523 (best: 0.8531)
Epoch 451/1000 - DiceMetric: 0.8519 (best: 0.8531)
...
Epoch 500/1000 - DiceMetric: 0.8518 (best: 0.8531)
Early stopping triggered - no improvement for 50 epochs
Restoring best model from epoch 450
Training completed at epoch 500 (early stopped)
```

### Required
- `enabled`: Enable/disable early stopping
- `monitor`: Metric name to monitor (must be in metrics list)

### Optional
- `patience`: Number of epochs to wait (default: 50)
- `min_delta`: Minimum improvement threshold (default: 0.0001)
- `mode`: 'max' for metrics (higher is better) or 'min' for losses (default: 'max')
- `restore_best_weights`: Load best checkpoint on stop (default: true)
- `baseline`: Initial value that must be exceeded (default: None)
- `start_epoch`: Epoch to start monitoring (default: 0)

## Additional Features (Future)

- **Multiple metrics**: Monitor combination of metrics (e.g., DiceMetric AND loss)
- **Divergence detection**: Stop if metric diverges/explodes
- **Learning rate reduction**: Reduce LR on plateau before stopping
- **Warmup**: Disable early stopping for initial epochs

## Related

- Complements existing checkpoint saving (saves best model already)
- Works with validation interval (`val_interval`)
- Integrates with existing metrics system

## References

- PyTorch Ignite: [`EarlyStopping` handler](https://pytorch.org/ignite/generated/ignite.handlers.early_stopping.EarlyStopping.html)
- Keras: [Early Stopping callback](https://keras.io/api/callbacks/early_stopping/)
- nnU-Net: Trains for fixed epochs, no built-in early stopping
